[water_heater] (1/4)  Implement API/Core/component for new water_heater component

dhoeben opened the pull request:
needs
- [x] https://github.com/esphome/esphome/pull/12585 (to avoid protobuf conflicts)
- [x] https://github.com/esphome/esphome/pull/12599

# What does this implement/fix?
Now split-up config for a new `water_heater` component. 

**This requires more pull requests!**

Order of PR:
1: This one
2: [water_heater template](https://github.com/esphome/esphome/pull/12516)
3: [water_heater web_server](https://github.com/esphome/esphome/pull/12511)
4: [water_heater tests](https://github.com/esphome/esphome/pull/12517)

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes: https://github.com/esphome/feature-requests/issues/2047)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
- aioesphomeapi: https://github.com/esphome/aioesphomeapi/pull/1459
- docs: https://github.com/esphome/esphome-docs/pull/5785
- HA-docs: https://github.com/home-assistant/home-assistant.io/pull/42599
- HA-core: https://github.com/home-assistant/core/pull/159201

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
water_heater:
  - platform: template
    id: my_boiler
    name: "Boiler"
    optimistic: true

    supported_modes:
      - eco
      - gas

    current_temperature: !lambda "return id(dhw_temp).state;"
    mode: !lambda "return id(current_mode);"

    visual:
      min_temperature: 10.0
      max_temperature: 85.0
      target_temperature_step: 0.5

    set_action:
      - lambda: |-
          ESP_LOGI("boiler", "New mode: %d", id(my_boiler).get_mode());
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

dhoeben committed:
Initial creation of water_heater in core

dhoeben committed:
Created template component

dhoeben committed:
Running lint scripts

dhoeben committed:
Running lint cpp scripts

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

dhoeben committed:
fixed issue

dhoeben committed:
Run scripts

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

dhoeben committed:
Update config to accept raw values as well

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

dhoeben committed:
Removed duplicate

dhoeben committed:
Removed duplicate

dhoeben committed:
Update issues

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

bdraco commented on esphome/components/api/api.proto:
```diff
@@ -1101,6 +1101,50 @@ message ClimateCommandRequest {
+
+message ListEntitiesWaterHeaterResponse {
+  option (id) = 132;
+  option (source) = SOURCE_SERVER;
```
check   option (base_class) = "InfoResponseProtoMessage";

bdraco commented on esphome/components/template/water_heater/template_water_heater.cpp:
```diff
@@ -0,0 +1,99 @@
+#include <set>
+
+namespace esphome {
+namespace template_ {
```
```suggestion
namespace esphome::template_ { 
```

bdraco commented on not correct, ESPHome uses clang-format to format:
```diff
@@ -0,0 +1,24 @@
+[1mdiff --git a/esphome/components/template/water_heater/__init__.py b/esphome/components/template/water_heater/__init__.py[m
```
Looks like accidental commit

bdraco commented on need to be changed::
```diff
@@ -0,0 +1,314 @@
+ DownArrow ...................... ESC-j ... Retrieve next command line.
+ TAB ...................................... Complete filename & cycle.
+ SHIFT-TAB ...................... ESC-TAB   Complete filename & reverse cycle.
+ ctrl-L ................................... Complete filename, list all.
```
Looks like accidental commit

bdraco commented on esphome/components/template/water_heater/automation.h:
```diff
@@ -0,0 +1,31 @@
+
+template<typename... Ts>
+class TemplateWaterHeaterPublishAction : public Action<Ts...>, public Parented<TemplateWaterHeater> {
+ public:
```
Please keep PR minimal, automations can be added later

dhoeben replied:
Why not implement it good all in once? It is all part of the component...

bdraco replied:
https://developers.esphome.io/contributing/submitting-your-work/

> Keep PRs as small and as focused as possible. Smaller PRs tend to be easier to understand and take less time to review. Large PRs (many hundreds or thousands of lines) by their nature (of being large) tend to keep changing which means reviewers have to revisit them over and over as they evolve. This isn't a tenable practice for project maintainers. Break your work into multiple, smaller PRs and link these PRs together with comments in the description so reviewers can follow the work more easily.

dhoeben replied:
The issue is that most of the commit lines are part of the protobuf of the api. So no matter how, this will be a big PR. It's a full integration.

bdraco replied:
All the more reason to keep it as small as possible.

dhoeben committed:
Update esphome/components/template/water_heater/template_water_heater.cpp

Co-authored-by: J. Nick Koston <nick+github@koston.org>

dhoeben committed:
remove accidental commit

dhoeben committed:
Merge branch 'dev' into water_heater

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

dhoeben committed:
Fix issues

dhoeben committed:
Fix issues

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

dhoeben commented:
I understand, I really do. But none of the checks will pass because it is all related to each other. Hence why I put it all in one bigger PR. I see what I can split up more tomorrow.

dhoeben committed:
Removed dependecies.lock

dhoeben committed:
Splitting up into smaller PR

dhoeben committed:
Add water heater component

dhoeben commented:
Now split up the config:

1: [water_heater api/core](https://github.com/esphome/esphome/pull/12498)
2: [water_heater component](https://github.com/esphome/esphome/pull/12515)
3: [water_heater template](https://github.com/esphome/esphome/pull/12516)
4: [water_heater web_server](https://github.com/esphome/esphome/pull/12511)

dhoeben committed:
Merge branch 'dev' into water_heater

dhoeben committed:
add codeowner

dhoeben committed:
Ran build_codeowners.py

dhoeben committed:
Merge pull request #1 from dhoeben/wh_component

Combine step 1 & 2

dhoeben committed:
Merge branch 'dev' into water_heater

bdraco committed:
use  FiniteSetMask

bdraco committed:
use NAN for sentinal value

bdraco committed:
use c++17 style namespaces

bdraco committed:
protobuf cleanup, missing ifdefs

bdraco committed:
use base classes for proto messages

bdraco committed:
dry

bdraco commented on esphome/components/water_heater/__init__.py:
```diff
@@ -0,0 +1,75 @@
+).extend(cv.COMPONENT_SCHEMA)
+
+
+async def setup_water_heater_core_(var, config):
```
missing typing

bdraco commented on esphome/components/water_heater/__init__.py:
```diff
@@ -0,0 +1,75 @@
+    if (max_temp := visual.get(CONF_MAX_TEMPERATURE)) is not None:
+        cg.add(var.set_visual_max_temperature_override(max_temp))
+
+
```
missing typing

bdraco committed:
stub for web_server

bdraco commented on esphome/components/water_heater/__init__.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@coroutine_with_priority(CoroPriority.CORE)
+async def to_code(config):
```
missing typing

bdraco committed:
encap, fixes

bdraco committed:
match schema to other components

bdraco committed:
missing dupe validator

bdraco committed:
Merge branch 'dev' into water_heater

bdraco committed:
away mode

bdraco committed:
remove legacy

bdraco committed:
persist state

bdraco committed:
fix prefs restore

bdraco committed:
current_temperature_step does not exist

https://developers.home-assistant.io/docs/core/entity/water-heater?_highlight=water

bdraco committed:
better align with climate

bdraco committed:
align fields

bdraco committed:
bot review comments

bdraco commented on esphome/components/water_heater/water_heater.h:
```diff
@@ -0,0 +1,221 @@
+
+void call_water_heater_update(WaterHeater *a);
+void register_water_heater(WaterHeater *a);
+
```
65e221f7e4

bdraco commented on esphome/components/web_server/list_entities.cpp:
```diff
@@ -135,6 +135,13 @@ bool ListEntitiesIterator::on_alarm_control_panel(alarm_control_panel::AlarmCont
 
+#ifdef USE_WATER_HEATER
+bool ListEntitiesIterator::on_water_heater(water_heater::WaterHeater *obj) {
+  // TODO: Implement water_heater web_server support
```
65e221f7e4

bdraco commented on esphome/components/water_heater/water_heater.cpp:
```diff
@@ -0,0 +1,211 @@
+
+WaterHeaterCall &WaterHeaterCall::to_call(WaterHeater *water_heater) {
+  water_heater->make_call().set_from_restore(*this).perform();
+  return *this;
```
65e221f7e4

bdraco commented on esphome/components/water_heater/water_heater.cpp:
```diff
@@ -0,0 +1,211 @@
+  return *this;
+}
+
+void WaterHeaterCall::apply(WaterHeater *water_heater) { *this = water_heater->make_call(); }
```
65e221f7e4

bdraco commented on esphome/components/water_heater/water_heater.cpp:
```diff
@@ -0,0 +1,211 @@
+      if (this->target_temperature_ > traits.get_max_temperature())
+        this->target_temperature_ = traits.get_max_temperature();
+    }
+  }
```
65e221f7e4

bdraco committed:
add missing two point support, on/off

bdraco committed:
collapse duplicate fields

bdraco commented:
This should be good to merge right after we have the aioesphomeapi with the same protobuf changes + api support bump merged in

dhoeben committed:
Merge branch 'dev' into water_heater

dhoeben committed:
Add template options to visual.min/max_temperature

dhoeben committed:
Removed dependicies.lock

dhoeben committed:
Merge branch 'dev' into water_heater

bdraco reviewed (changes requested):
Hey @dhoeben,

Thanks for all the work on this water_heater component! I noticed some changes were added after the PR was already approved, specifically making `CONF_MIN_TEMPERATURE` and `CONF_MAX_TEMPERATURE` templatable with `TemplatableValue<float>`.

This deviates from how the climate component handles visual temperature overrides. Climate uses plain `float` values:

```python
# climate/__init__.py
cv.Optional(CONF_MIN_TEMPERATURE): cv.temperature,  # plain float
cv.Optional(CONF_MAX_TEMPERATURE): cv.temperature,  # plain float
```

```cpp
// climate.h
void set_visual_min_temperature_override(float visual_min_temperature_override);
void set_visual_max_temperature_override(float visual_max_temperature_override);
```

I have family in town for the holidays and don't have the bandwidth to do another full review of scope-creep changes right now. Could you please revert the templatable additions so we can get this PR merged? We can discuss adding templatable visual overrides as a separate enhancement in a future PR—that way we can properly evaluate the use case and potentially add it to both climate and water_heater together for consistency.

To summarize what needs to be reverted:
- Change `cv.templatable(cv.temperature)` back to `cv.temperature` for min/max in the visual schema
- Change `TemplatableValue<float>` back to `float` for the visual override members
- Remove the `#include "esphome/core/automation.h"` if it was only added for TemplatableValue
- Update the setters to take plain `float` instead of `const TemplatableValue<float> &`

Thanks for understanding, and happy holidays!

dhoeben committed:
revert changes

bdraco reviewed (approved):
Thanks @dhoeben
